### ISSUE TITLE
Test osmdbt produced diffs

### ISSUE THREAD

**Issue by pnorman** [openstreetmap/operations#475]
As part of #154 we've got new diffs produced by osmdbt at https://planet.openstreetmap.org/replication/test

We need to get this tested by users.

The first thing to do is put out a call for people to test on the dev@ list.

**Comment by mmd-osm** [comment]
Two questions, as I don't know if this is intentional.

1. Sequence numbers in https://planet.openstreetmap.org/replication/test seem to be out of sync with the current replication directory. Previously, diffs were created at a fixed time (at zero seconds each minute). Now it looks like the duration of one create-diff run determines the start time of the next job. This results will be fewer but larger files overall.

2. I don't see the `--max-changes` parameter in place for create-diff (https://github.com/openstreetmap/osmdbt/blob/master/man/osmdbt-create-diff.md#options). See https://github.com/openstreetmap/osmdbt/issues/10 for discussion.

**Comment by tomhughes** [comment]
Well I wouldn't expect sequence numbers to be in sync - to start with there was a delay at the start between when I took a snapshot of the current number and enabled replication and when I actually started the cron job so for the first few minutes there are only a few manually run diffs.

After that there is no guarantee that they run at the same boundaries or even exactly at the same frequency - both attempt to run once a minute and generate one diff each time but there was never been any guarantee of that.

All of that is why I say it's a dead end because when we switch over properly we will restart from whatever point we are then at.

I don't think it would be safe to use `--max-changes` as if we ever started to produce more than that number of changes on average per minute we would just start to slip further and further behind. I don't think that can happen with osmosis as a single run will, if necessary, output more than one file until it has caught up.

**Comment by mmd-osm** [comment]
Regarding the first point: I really meant that both sequenceNumbers keep diverging more and more. It started with 9 minutes of a difference (which is expected), and turned into a few hours in the meantime. I was wondering if that's the way it's supposed to be.

Regarding the second point: that's meant as a safeguard in case of several hours of outage. In fact, osmosis doesn't create multiple files in a single run, as my example in https://github.com/openstreetmap/osmdbt/issues/10 shows.

**Comment by tomhughes** [comment]
Maybe it's only the merging task that produces the hourly/daily diffs that can output multiple files. I'm pretty sure that does do so or we would have had problem with it at various points.

In any case I don't believe the osmosis minutely replicator has any size limit.

I'm not really bothered about producing them exactly on zero seconds, just that they are produced at roughly regular intervals. Currently I believe it's set to start one minute after the previous run ends. I could change that to be based on the start time of the previous run but that would be risky if a run ever took more than one minute.

**Comment by mmd-osm** [comment]
> In any case I don't believe the osmosis minutely replicator has any size limit.

It does in fact have a limit, although it's only implicit based some on the implementation detail and nothing you could configure using some parameter setting. That's why we introduced max-limit to have the same behaviour in place.

> Currently I believe it's set to start one minute after the previous run ends. I could change that to be based on the start time of the previous run but that would be risky if a run ever took more than one minute.

osmdbt would find out if a previous run is still active, so that shouldn't be a concern, after all.

**Comment by tomhughes** [comment]
Well what value are you suggesting, because I don't see anywhere in this ticket or the referenced one where you suggest a value, or even explain what counts as a "change" for the purposes of that switch. Is it the total number of object changes? Does that include changset creations/closes?

**Comment by mmd-osm** [comment]
`--max-changes` represents the total number of object changes plus one extra line for each database commit. If a user uploads multiple times using the the same changeset number, you would see one commit line for each upload. Maybe that could be made a bit more explicit in the documentation.

We actually left determining a good value to the testing phase. I would use 500'000 as a starting point, and monitor osc.gz. file sizes afterwards. Unfortunately, that would mean to stop the replication process for a few hours and then restart it. During normal operation, you would probably never hit that limit.

**Comment by tomhughes** [comment]
I would suggest something about an order of magnitude less than that to get similar results to the current system. Looking at one of the diffs from that example you gave there are about 53000 object changes in it and about 550 changesets.

**Comment by tomhughes** [comment]
I was wrong about the timer by the way - it is using the activation time as the start point for the next execution.

I suspect what you're seeing is the result of the default `AccuracySec` setting on systemd timers that will try and adjust the time to coalesce timers. It default to a minute which might be a bit high in this case. That said it doesn't seem to cause much delay in general - this is the recent executions:

```
2020-11-20T12:10:48.169581+0000 ironbelly.openstreetmap.org systemd[1]: Started Minutely replication.
2020-11-20T12:11:48.179337+0000 ironbelly.openstreetmap.org systemd[1]: Started Minutely replication.
2020-11-20T12:12:48.186558+0000 ironbelly.openstreetmap.org systemd[1]: Started Minutely replication.
2020-11-20T12:13:48.196881+0000 ironbelly.openstreetmap.org systemd[1]: Started Minutely replication.
2020-11-20T12:14:48.208463+0000 ironbelly.openstreetmap.org systemd[1]: Started Minutely replication.
2020-11-20T12:15:48.220148+0000 ironbelly.openstreetmap.org systemd[1]: Started Minutely replication.
2020-11-20T12:16:48.231075+0000 ironbelly.openstreetmap.org systemd[1]: Started Minutely replication.
2020-11-20T12:18:01.311924+0000 ironbelly.openstreetmap.org systemd[1]: Started Minutely replication.
```

**Comment by tomhughes** [comment]
Ha and then the last one there does show more of a jump...

**Comment by joto** [comment]
Regarding timer: `osmdbt-get-log` and `osmdbt-create-diff` were designed in a way that they can run independently. `osmdbt-get-log` doesn't do much so there should never be the problem that it runs so long it will overlap with itself. So we could run it exactly once per minute. Then the `osmdbt-create-diff` can run independently in a loop and take however long it needs. With normal operation, both should run exactly once per minute, if `osmdbt-create-diff` is a bit slow, it will pick up however many log files `osmdbt-get-log` has created in the mean time, "join" them and create one diff file. So as a result we should get exactly minutely diffs if everything runs well and, if something is slow, less then minutely diffs with more data in them (up to the `max-changes`).
`osmdbt-create-diff` uses a lock file to make sure it doesn't run twice. 

So we would have two scripts running.

This once per minute:
```
osmdbt-catchup --quiet
osmdbt-get-log --quiet --catchup
```

This one should pick up the log files generated from the first script as fast as possible. Maybe something like:

```
while true; do
   osmdbt-create-diff --quiet --max-changes=50000
   sleep 5
done
```

Not sure what would be best here.

**Comment by mmd-osm** [comment]
I had a quick chat with @Firefishy about our options to replace osmosis for hourly and daily diff generation (see [link](https://github.com/openstreetmap/chef/blob/master/cookbooks/planet/recipes/replication.rb#L349)).  I believe osmium tool would be a good fit, in particular https://docs.osmcode.org/osmium/latest/osmium-cat.html:

```
Concatenate all change files in the ‘changes’ directory into one:
osmium cat -o all-changes.osc.gz changes/*.osc.gz
```

I'm not exactly sure if this will produce exactly the same result as osmosis does today, and if it could be used as a drop-in replacement.

**Comment by joto** [comment]
No, `osmium cat` is the wrong command for that, what you are looking for is `osmium merge-changes`.

**Comment by mmd-osm** [comment]
Ah, yes, you're right, we need to sort objects, rather than just concatenate files. In that case we would run `osmium merge-changes` without the simplify parameter, as we must keep all intermediate object versions.

Do you have some experience from previous projects how the state.txt could be updated as part of that process, and which .osc.gz would be included in a merge-changes run?

**Comment by tomhughes** [comment]
That's the basic tool yes, what's needed is something around it to gather an hour or day's worth of changes and feed them to the merger. I did look and I don't think there's anything obvious at the moment among what already exists.

It probably just needs a bit of work to write a new pyosmium script to fetch the outstanding minutely and merge them into hourlies and then merge hourlies into dailies.

Basically at the moment the osmosis configuration we use for hourlies fetches up to ten hours worth of minutleies on each run and spits that out in hourly batches, and we run it four times an hour so if it gets behind it catches up at a 40x rate roughly and then similar for dailies.

**Comment by lonvia** [comment]
pyosmium-get-changes can do the non-simplified merging for you already. It uses a size parameter to cap the number of changes merged in one pass. I can also add a time constraint if you prefer to keep the existing behaviour. If I understand you right, then you would also need a 'minimum time' parameter where you can state something along the line: only download if there is at least 50min of changes since the last run. Aka we'd end up running something like:

```
pyosmium-get-changes --server https://minutelies --min-period 50m --max-period 10h -f minutely.sequence --no-deduplicate $HOUR_SEQ.osc.gz
```
every 15 minutes and then save the diff away etc. when pyosmium is successful. Would that be about right?

**Comment by tomhughes** [comment]
Yes minimum and maximum time constraints would be good.

I think there's already an exit status that indicates if the limit was hit right? So we can just loop until we don't hit the limit (or don't produce anything because there isn't 50 minutes worth).

**Comment by lonvia** [comment]
Yes, the exit status allows that. The update script for the renderers should be doing something similar already.

Another thought: who produces the state files at the moment? In particular, I wonder how the timestamp information is created. I suspect that you would need to get the timestamp information from the state file of the source. That would be something else that needs adding to pyosmium-get-changes then.

**Comment by tomhughes** [comment]
The state files are produced by osmosis.

The timestamps it output don't seem to correspond to the last minutely it consumed or even to the latest timestamp in the diff but in fact appear to be exactly aligned to the hour so maybe I misunderstood how it works and it actually collects stuff up to the hour boundary and merges it? this is the configuration it uses for hourly:

```
# The URL of the directory containing change files.
baseUrl=file:///store/planet/replication/minute

# The length of an extraction interval in seconds (3600 = 1 hour).
intervalLength=3600

# Defines the maximum time interval in seconds to download in a single invocation.
# Setting to 0 disables this feature.
maxInterval=36000
```

That is used as the configuration for `--merge-replication-files`.

**Comment by pnorman** [comment]
Now in production. Any problems can be ticked as new issues.

**Comment by mmd-osm** [comment]
Should we take the hourly and daily diff topic to fully retire osmosis to a new issue?
